documentation/specs/src/interoperability/ethereum-bridge/ethereum_smart_contracts.md

```diff
@@ -1,27 +1,182 @@
-2. Look up the validator set corresponding to said epoch.
-3. Verify that the signatures included amount to at least 2 / 3 of the
-   total stake.
-4. Check the validity of each signature.
-
-If all the above verifications succeed, the contract may affect the
-appropriate state change, emit logs, etc.
\ No newline at end of file
+## Contracts
+
+There are five smart contracts that make up an Ethereum bridge deployment.
+
+- Proxy
+- Bridge
+- Governance
+- Vault
+- wNAM
+
+### Proxy
+
+The _Proxy_ contract serves as a dumb storage for holding the addresses of other
+contracts, specifically the _Governance_ contract, the _Vault_ contract and the
+current _Bridge_ contract. Once deployed, it is modifiable only by the
+_Governance_ contract, to update the address for which contract is the current
+_Bridge_ contract.
+
+The _Proxy_ contract is fixed forever once the bridge has been deployed.
+
+### Bridge
+
+The _Bridge_ contract is the only contract that unprivileged users of the bridge
+may interact with. It provides methods for transferring ERC20s to Namada
+(holding them in escrow in the _Vault_), as well as releasing escrowed ERC20s
+from the _Vault_ for transfers made from Namada to Ethereum. It holds a
+whitelist of ERC20s that may cross the bridge, and this whitelist may be updated
+by the _Governance_ contract.
+
+### Governance
+
+The _Governance_ contract may "upgrade" the bridge by updating the _Proxy_
+contract to point to a new _Bridge_ contract and/or a new _Governance_ contract.
+It may also withdraw all funds from the _Vault_ to any specified Ethereum
+address, if a quorum of validators choose to do so.
+
+### wNAM
+
+The _wNAM_ contract is a simple ERC20 token with a fixed supply, which is all
+minted when the bridge is first deployed. After initial deployment, the entire
+supply of _wNAM_ belongs to the _Vault_ contract. As NAM is transferred from
+Namada to Ethereum, wNAM may be released from the _Vault_ by the _Bridge_.
+
+The _wNAM_ contract is fixed forever once the bridge has been deployed.
+
+### Vault
+
+The _Vault_ contract holds in escrow any ERC20 tokens that have been sent over
+the bridge to Namada, as well as a supply of _wNAM_ ERC20s to represent NAM that
+has been sent from Namada to Ethereum. Funds held by the _Vault_ may only be
+spendable by the current _Bridge_ contract. When ERC20 tokens are transferred
+from Ethereum to Namada, they must be deposited to the _Vault_ via the _Bridge_
+contract.
+
+The _Vault_ contract is fixed forever once the bridge has been deployed.
+
+## Namada-side configuration
+
+When an account on Namada becomes a validator, they must provide two Ethereum
+secp256k1 keys:
+
+- the bridge key - a hot key for normal operations
+- the governance key - a cold key for exceptional operations, like emergency
+  withdrawal from the bridge
+
+These keys are used to control the bridge smart contracts, via signing of
+messages. Validators may be challenged periodically to prove they still retain
```

Comment:
The hot keys don't need a challenge I guess since once an epoch they are used to sign over valset updates. Of course that only tests a quorum of validators. For cold keys, this is an interesting point.

Reply:
if we use that strategy we discussed some 2 weeks ago or so, to prove ownership of all keys, then you also solve the problem of proving ownership of eth cold keys.

https://excalidraw.com/#json=GzskiES12LxvLl-7Y_wGk,qPn9jmoqX2VUnMbJSm86hQ